Manage the subsumption check

feliciahalim opened the pull request:
This PR currently has 3 main changes:

1. Change the subsumption check iterator into reverse iterator. From the observation of `klee-examples/Regexp.c `and 2 `coreutils` program. The successful subsumption check mostly occur on the newest entry. Therefore, to speed up finding the suitable entry for subsumption check, we can start searching from the end of the list where the newest entry is located.

2. Exhaustive search for all table entry on particular node for finding subsumption check may costly especially after studying the case of `klee-examples/Regexp` where the subsumption only occur mostly on the first, second, or third newest entry of the table. Therefore, I add the user option that enable user to specify the number of maximum fail subsumption (`max-fail-subsume`).

3. I also add another user option to specify the subsumption table size (`subsumption-table-size`) that can be used together with max-fail-subsume option. If we already set max-fail-subsume, so we might not need to store all of the entries since it may not be checked anyway. For example: if the max-fail-subsume equals to 10, we might interested to store only the 10 newest entries because the rest of entries are not checked anyway, so we can reduce the space consumption. 

Below are the experiment result for `klee-examples/Regexp.c` `-max-fail-subsume=10 -subsumption-table-size=10
`
```
KLEE: done: total instructions = 1001434
KLEE: done: completed paths = 2547, among which
KLEE: done:     early-terminating paths (instruction time limit, solver timeout, max-depth reached) = 0
KLEE: done:     subsumed paths = 1052
KLEE: done:     error paths = 237
KLEE: done:     program exit paths = 1258
KLEE: done: generated tests = 2312, among which
KLEE: done:     early-terminating tests (instruction time limit, solver timeout, max-depth reached) = 0
KLEE: done:     subsumed tests = 1052
KLEE: done:     error tests = 2
KLEE: done:     program exit tests = 1258
Execution Time : 42.33
```
compare to result from`No option
`
```
KLEE: done: total instructions = 986871
KLEE: done: completed paths = 2538, among which
KLEE: done:     early-terminating paths (instruction time limit, solver timeout, max-depth reached) = 0
KLEE: done:     subsumed paths = 1072
KLEE: done:     error paths = 237
KLEE: done:     program exit paths = 1229
KLEE: done: generated tests = 2303, among which
KLEE: done:     early-terminating tests (instruction time limit, solver timeout, max-depth reached) = 0
KLEE: done:     subsumed tests = 1072
KLEE: done:     error tests = 2
KLEE: done:     program exit tests = 1229
Execution time: 93.63
```

We might lose few numbers of subsumption, but exhaustive search to all of entries of subsumption table entry are also expensive and also we could give user the flexibility to experiment. With specifying `max-fail-subsume` and `subsumption-table-size`, the execution time is changed from 93.63 to 42.33.

domainexpert commented:
@feliciahalim I think both `-max-fail-subsume` and `-subsumption-table-size` can be unified into one, say just `-subsumption-table-size`, which determines the size of the table, which is implemented using FIFO queue, so whenever the size limit is reached, the oldest entry is thrown away.

feliciahalim commented:
@domainexpert Thanks for the suggestion. But I think FIFO queue may not be suitable for doing iteration on each element and particularly used for accessing its front element.  I actually plan to remove `-subsumption-table-size` option because we use list and this option may requires remove elements and shifting` n-1` element to the left every time we inserted a new element and capacity is full. 
For example : capacity is 4
`( a b c d)  -> insert e -> ( a b c d e) -> (b c d e)`
I also think this time we may keep the all of the entries first because the option may only favorable for DFS, but not sure with other search heuristics.

domainexpert commented:
@feliciahalim Your diagram depicts FIFO queue exactly as I meant it. In C, you can implement it using doubly-linked list, where you keep the pointers to the first and last elements. A new element can be added to before the first element, and the last element removed, traversing the list backward such that you assign the new last element to the previously second-to-last element.

feliciahalim commented:
@domainexpert  I have integrated `max-fail-subsume` and `-subsumption-table-size` as suggested. I implemented with `std::deque` that allow access element from both front or back, therefore I can remove the oldest element with `pop_front() `when table size increased more than the specified  `max-fail-subsume`. 

This is the result when using `-max-fail-subsume = 3` for Regexp.c
```
KLEE: done: total instructions = 992837
KLEE: done: completed paths = 2546, among which
KLEE: done:     early-terminating paths (instruction time limit, solver timeout, max-depth reached) = 0
KLEE: done:     subsumed paths = 1059
KLEE: done:     error paths = 237
KLEE: done:     program exit paths = 1250
KLEE: done: generated tests = 2311, among which
KLEE: done:     early-terminating tests (instruction time limit, solver timeout, max-depth reached) = 0
KLEE: done:     subsumed tests = 1059
KLEE: done:     error tests = 2
KLEE: done:     program exit tests = 1250
Execution Time: 43.23
```
When no option is specified:
```
KLEE: done: total instructions = 986871
KLEE: done: completed paths = 2538, among which
KLEE: done:     early-terminating paths (instruction time limit, solver timeout, max-depth reached) = 0
KLEE: done:     subsumed paths = 1072
KLEE: done:     error paths = 237
KLEE: done:     program exit paths = 1229
KLEE: done: generated tests = 2303, among which
KLEE: done:     early-terminating tests (instruction time limit, solver timeout, max-depth reached) = 0
KLEE: done:     subsumed tests = 1072
KLEE: done:     error tests = 2
KLEE: done:     program exit tests = 1229
Execution Time: 94.38
```

domainexpert commented:
@feliciahalim In the help message for the max-fail-subsume option:
1. Please remove the first clause containing "flexibility" as it does not say anything. That is, instead of starting with "This option ...," I recommend starting from "To set the maximum ..."
2. "the number of subsumption table size" should be "the number of subsumption table entries."
3. Please explain what the default=0 means. At present it reads as if by default, the value is 0, and therefore no entry can be stored into the table by default.

feliciahalim commented:
@domainexpert I have updated the description of  `max-fail-subsume `option as suggested and added (default = 0) with (off), which follows MaxCoreSolverTime description.

domainexpert commented:
@feliciahalim Thanks! Just a couple of small things: In the same description:
1. fail subsumption check -> failed subsumption check
2. this options is selected -> this option is specified
3. Remove the clause "the number of subsumption table entries is equal to the number of max-fail-subsume." and conjoin "When ..." with the preceding statement, replacing "When" with and.
4. "the number of entries are more than specified max-fail-subsume" -> "the number of subsumption table entries is more than the specified value"

feliciahalim commented:
@domainexpert I have updated the description as suggested. Thank you.
